File: problems/934_shortest_bridge/solution.cpp
#include <vector>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <algorithm>
#include <climits>
#include <queue>
#include <stack>
#include <cmath>
#include <numeric>
#include <set>
#include <map>
using namespace std;

class Solution {
public:
    int shortestBridge(vector<vector<int>>& grid) {
        int n = grid.size();
        vector<vector<bool>> visited(n, vector<bool>(n, false));
        queue<pair<int,int>> q;
        bool found = false;
        for (int i = 0; i < n && !found; i++)
            for (int j = 0; j < n && !found; j++)
                if (grid[i][j] == 1) { dfs(grid, visited, q, i, j, n); found = true; }
        int steps = 0;
        int dirs[4][2] = {{0,1},{0,-1},{1,0},{-1,0}};
        while (!q.empty()) {
            int sz = q.size();
            for (int k = 0; k < sz; k++) {
                auto [x, y] = q.front(); q.pop();
                for (auto& d : dirs) {
                    int nx = x + d[0], ny = y + d[1];
                    if (nx >= 0 && nx < n && ny >= 0 && ny < n && !visited[nx][ny]) {
                        if (grid[nx][ny] == 1) return steps;
                        visited[nx][ny] = true;
                        q.push({nx, ny});
                    }
                }
            }
            steps++;
        }
        return -1;
    }
    void dfs(vector<vector<int>>& grid, vector<vector<bool>>& visited, queue<pair<int,int>>& q, int i, int j, int n) {
        if (i < 0 || i >= n || j < 0 || j >= n || visited[i][j] || grid[i][j] == 0) return;
        visited[i][j] = true;
        q.push({i, j});
        dfs(grid, visited, q, i+1, j, n);
        dfs(grid, visited, q, i-1, j, n);
        dfs(grid, visited, q, i, j+1, n);
        dfs(grid, visited, q, i, j-1, n);
    }
};

File: problems/934_shortest_bridge/solution.py
from typing import List


class Solution:
    def shortestBridge(self, grid: List[List[int]]) -> int:
        n = len(grid)
        visited = [[False] * n for _ in range(n)]
        queue: deque = deque()
        found = False
        for i in range(n):
            if found:
                break
            for j in range(n):
                if grid[i][j] == 1:
                    self._dfs(grid, visited, queue, i, j, n)
                    found = True
                    break
        steps = 0
        dirs = [(0, 1), (0, -1), (1, 0), (-1, 0)]
        while queue:
            for _ in range(len(queue)):
                x, y = queue.popleft()
                for dx, dy in dirs:
                    nx, ny = x + dx, y + dy
                    if 0 <= nx < n and 0 <= ny < n and not visited[nx][ny]:
                        if grid[nx][ny] == 1:
                            return steps
                        visited[nx][ny] = True
                        queue.append((nx, ny))
            steps += 1
        return -1

    def _dfs(self, grid: List[List[int]], visited: List[List[bool]], queue: deque, i: int, j: int, n: int) -> None:
        if i < 0 or i >= n or j < 0 or j >= n or visited[i][j] or grid[i][j] == 0:
            return
        visited[i][j] = True
        queue.append((i, j))
        self._dfs(grid, visited, queue, i + 1, j, n)
        self._dfs(grid, visited, queue, i - 1, j, n)
        self._dfs(grid, visited, queue, i, j + 1, n)
        self._dfs(grid, visited, queue, i, j - 1, n)
